src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,539 @@
+        {
+            var username = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
+            var password = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
+            var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+
+            return awsCredentials;
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEcsResponse()
+        {
+            var relativeUri = Environment.GetEnvironmentVariable("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI");
+
+            if (relativeUri == null)
+            {
+                return null;
+            }
+
+            var response = AwsHttpClientHelper.GetECSResponseAsync(relativeUri).GetAwaiter().GetResult();
+            var parsedResponse = BsonDocument.Parse(response);
+            var username = parsedResponse.GetValue("AccessKeyId", null)?.AsString;
+            var password = parsedResponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedResponse.GetValue("Token", null)?.AsString;
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+
+            return awsCredentials;
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEc2Response()
+        {
+            var response = AwsHttpClientHelper.GetEC2ResponseAsync().GetAwaiter().GetResult();
+            var parsedResponse = BsonDocument.Parse(response);
+            var username = parsedResponse.GetValue("AccessKeyId", null)?.AsString;
+            var password = parsedResponse.GetValue("SecretAccessKey", null)?.AsString;
+            var sessionToken = parsedResponse.GetValue("Token", null)?.AsString;
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = ToSecureString(password),
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+
+            return awsCredentials;
+        }
+
+        private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
```

Comment:
Refactor this method to:

    private static string ExtractSessionTokenFromMechanismProperties(IEnumerable<KeyValuePair<string, string>> properties)
    {
        if (properties != null)
        {
            foreach (var pair in properties)
            {
                if (pair.Key.ToUpperInvariant() == "AWS_SESSION_TOKEN")
                {
                    return pair.Value;
                }
            }
        }

        return null;
    }

I think it is simpler to have only two return statements instead of three.